hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -516,59 +520,51 @@ def fixate_shrink_passes(self, passes):
-
-                # The choice of 3 is fairly arbitrary and was hand tuned
-                # to some particular examples. It is very unlikely that
-                # is the best choice in general, but it's not an
-                # unreasonable choice: Making it smaller than this would
-                # give too high a chance of an otherwise very worthwhile
-                # pass getting screened out too early if it got unlucky,
-                # and making it much larger than this would result in us
-                # spending too much time on bad passes.
-                max_failures = 3
+                if can_discard:
+                    can_discard = self.remove_discarded()
 
+                before_sp = self.shrink_target
+
+                # Run the shrink pass until it fails to make any progress
+                # ten times in a row. This implicitly boosts shrink passes
+                # that are more likely to work.
+                failures = 0
+                max_failures = 20
                 while failures < max_failures:
-                    prev_calls = self.calls
                     prev = self.shrink_target
-                    if sp.step():
-                        any_ran = True
-                    else:
+                    initial_calls = self.calls
+                    if not sp.step(random=failures >= max_failures // 2):
                         break
-                    if prev_calls != self.calls:
-                        if can_discard:
-                            can_discard = self.remove_discarded()
-                        if prev is self.shrink_target:
-                            failures += 1
-                        else:
-                            successes += 1
-                if successes > 0:
-                    successful_passes.add(sp)
-
-            # If only some of our shrink passes are doing anything useful
-            # then run all of those to a fixed point before running the
-            # full set. This is particularly important when an emergency
-            # shrink pass unlocks some non-emergency ones and it suddenly
-            # becomes very expensive to find a bunch of small changes.
-            if 0 < len(successful_passes) < len(passes):
-                self.fixate_shrink_passes(successful_passes)
+                    any_ran = True
+
+                    # Don't count steps that didn't actually try to do
+                    # anything as failures.
+                    if initial_calls == self.calls:
+                        continue
+                    if prev is not self.shrink_target:
+                        failures = 0
+                    else:
+                        failures += 1
+
+                if self.shrink_target is before_sp:
+                    reordering[sp] = 1
+                elif len(self.buffer) < len(before_sp.buffer):
+                    reordering[sp] = -1
+                else:
+                    reordering[sp] = 0
```

Comment:
These three cases could use quick comments clarifying that they give early priority to passes that succeeded at length-shrinking, medium priority to passes that succeeded at value-shrinking, and demote passes that didn't achieve anything.